src/python/pants/engine/README.md

```diff
@@ -75,52 +75,74 @@ In cases where this search detects any ambiguity (generally because there are tw
+
+The typical usage of `@dataclass` and `@dataclasstype` is to compose them to wrap a typed class definition:
+
+*Note that the [3rdparty `dataclasses` library](https://github.com/ericvsmith/dataclasses) must be in your BUILD file and `import`ed in Python 3.6!*
 
 ```python
-class FormattedInt(datatype(['content'])): pass
+from dataclasses import dataclass
+
+from pants.util.objects import dataclasstype
+
+
+# NB: @dataclasstype currently requires that `frozen=True` must be set in the `@dataclass`
+# decorator!
+@dataclasstype
+@dataclass(frozen=True)
+class FormattedInt:
+  content: str
 
 @rule(FormattedInt, [int])
-def int_to_str(an_int):
+def int_to_str_conversion(an_int):
   return FormattedInt('{}'.format(an_int))
 
-# Field values can be specified with positional and/or keyword arguments in the constructor:
+# Instances can be constructed and accessed the same way as for normal `@dataclass`es.
 x = FormattedInt('a string')
-x = FormattedInt(content='a string')
-
-# Field values can be accessed after construction by name or index:
+assert x == FormattedInt(content='a string')
 print(x.content)    # 'a string'
-print(x[0])         # 'a string'
 
 # datatype objects can be easily inspected:
-print(x)            # 'FormattedInt(content=a string)'
+print(repr(x))                  # 'FormattedInt(content="a string")'
+print(str(x))                   # 'FormattedInt(content<Exactly(str)>=a string)'
+
+# Most mypy types, including parameterized types, may be used in field definitions.
+@dataclasstype
+@dataclass(frozen=True)
+class TypedDatatype:
+  """Example of a datatype with a more complex field type constraint."""
+  field_name: Union[str, int]
+
+print(str(TypedDatatype("huh"))) # 'TypedDatatype(field_name<Exactly(str or int)>=huh)'
 ```
 
-#### Types of Fields
+#### API Stability with `DatatypeMixin`
 
-`datatype()` accepts a list of *field declarations*, and returns a type which can be subclassed. A
-*field declaration* can just be a string (e.g. `'field_name'`), which is then used as the field
-name, as with `FormattedInt` above. A field can also be declared with a tuple of two elements: the
-field name string, and a `TypeConstraint` for the field (e.g. `('field_name',
-Exactly(FieldType))`). The bare type name (e.g. `FieldType`) can also be used as a shorthand for
-`Exactly(FieldType)`. If the tuple form is used, the constructor will create your object, then raise
-an error if the field value does not satisfy the type constraint.
+[`DatatypeMixin`](../util.objects.py) is intended to decouple the *declaration* of engine `Param` types from their *usage*. In particular, `DatatypeMixin` declares a few `@abstractmethod`s such as `get_type_dict()`, which it uses to implement several generic methods, including `__str__()`, `__repr__()`, and `copy()` (which returns a new instance of the same class, with one or more fields modified).
```

Comment:
Will it be feasible to get rid of `DatatypeMixin` and `@dataclasstype` at some point? It's a bit of a weird appendage. Any special methods on a datatype that the engine requires it can presumably implement as functions (e.g., `copy(dataclass_instance, **kwargs)` instead of `dataclass_instance.copy(**kwargs)`).

It seems best if the datatypes have little or no API at all, other than their declared fields.  And being able to say "your pants products can be any frozen dataclass, with no extra boilerplate or syntax" is pretty hot.

Reply:
The other conditions described in the `Requirements for an Engine Param` are pretty important: must have useful eq and a useful hash. If there is a way to guarantee that on a dataclass, then maybe.

Reply:
All dataclasses get a useful `__eq__` by default (using, effectively, tuple comparison of fields), unless you set `eq=False`. And they all get a __hash__ if `frozen=True` and `eq=True`. 

True, both of these being meaningful depend on the types of the fields, but we would still be validating those, so we can ensure that.

It would be nice if our `@datatype` was just an alias for `@dataclass(frozen=True)` (with `eq=True` by default), as long as we can use the latter as well.